### PR TITLE
Refactoring the content loading system

### PR DESCRIPTION
**Pull request by abosma**
Currently data is loaded immediately from /Content/PackageName, this was now changed to follow the structure on the league sandbox website: [(given package structure)](https://cdn.discordapp.com/attachments/261986574487519232/602913723945844745/unknown.png). Items are now also not hard-coded but checks inside of the given datapackage in GameMode.json.

Next thing is see if it's better to merge gamemode and data or to keep it like this, then do whatever is decided.

### CODE REVIEW

**Comment by abosma** [comment]
I implemented a recursive package loading system. GameMode is not used at all anymore, every package must now contain a packageInfo.json with a structure like: `{ "dependencies": [ "Self-PackageName", "Other-PackageName" ] }`. Every other packages' dependencies will be loaded in as well.

GameInfo.json does not have a gameMode line anymore under game but now a dataPackage line with the initial loaded data package as value.

**Comment by abosma** [comment]
Removed the need for Self-PackageName in packageInfo.json, the new json format is: `{ "dependencies": [ "Other-PackageName" ] }`

**Comment by abosma** [comment]
Data is now again loaded from /Content/PackageName, scripts that frequently get edited (Buffs and Champions) have now been moved to LeagueSandbox-Scripts. LeagueSandbox-Default now has a dependency on that folder.

Other than that some error handling was also added in case a package does not containt packageInfo.json, it will now skip the dependencies from that package but still load it in.

**Comment by abosma** [comment]
Loading scripts and buffs from another package does not work well yet, making a more robust version of content loading.

**Comment by abosma** [comment]
While currently loading in multiple packages and dependency chaining works, it might be better to look at creating a nicer version of the current content loading system. Maybe create package classes that handle loading in themselves.

**Comment by abosma** [comment]
Package class added. This is created per given PackageName in GameInfo and dependencies inside of this package. This package loads itself in, instead of ContentManager loading this data in. Loaded in content is now no longer in ContentManager but split up per Package, ContentManager holds these packages.

Things that are not yet loaded in by packages themselves are: Map spawns and NavGrid.

**Comment by abosma** [comment]
MapSpawns and NavGrids now get loaded in through the package class. The spawns and navgrid that is chosen is currently based on load order of the packages starting with the given package in GameInfo.json.

**Comment by abosma** [comment]
The added commands are !loadpackage {packagename} and !showpackages

File: Content/LeagueSandbox-Scripts/Buffs/NidaleeTransformedQBuff/NidaleeTransformedQBuff.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿using GameServerCore.Domain;
+using GameServerCore.Domain.GameObjects;
+using LeagueSandbox.GameServer.GameObjects;
+using LeagueSandbox.GameServer.Scripting.CSharp;
+
+namespace NidaleeTransformedQBuff
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/Buffs/NidaleeTransformedQBuff/NidaleeTransformedQBuff.cs]: Unnecessary addition.

File: Content/LeagueSandbox-Scripts/Buffs/NidaleeEBuff/NidaleeEBuff.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+﻿using GameServerCore.Domain;
+using GameServerCore.Domain.GameObjects;
+using GameServerCore.Enums;
+using LeagueSandbox.GameServer.GameObjects;
+using LeagueSandbox.GameServer.GameObjects.Stats;
+using LeagueSandbox.GameServer.Scripting.CSharp;
+using static LeagueSandbox.GameServer.API.ApiFunctionManager;
+
+namespace NidaleeEBuff
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/Buffs/NidaleeEBuff/NidaleeEBuff.cs]: Unnecessary, same as WBuff.

File: Content/LeagueSandbox-Scripts/Champions/Nidalee/E.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
+﻿using GameServerCore.Domain;
+using GameServerCore.Domain.GameObjects;
+using LeagueSandbox.GameServer.GameObjects.AttackableUnits.AI;
+using LeagueSandbox.GameServer.Scripting.CSharp;
+
+namespace Spells
+{
+    class PrimalSurge : IGameScript
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/Champions/Nidalee/E.cs]: Unnecessary, same as NidaleeE/WBuff.

File: Content/LeagueSandbox-Scripts/Champions/Nidalee/R.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
+﻿using GameServerCore.Domain;
+using GameServerCore.Domain.GameObjects;
+using LeagueSandbox.GameServer.Content;
+using LeagueSandbox.GameServer.Scripting.CSharp;
+using static LeagueSandbox.GameServer.API.ApiFunctionManager;
+
+namespace Spells
+{
+    class AspectOfTheCougar : IGameScript
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/Champions/Nidalee/R.cs]: Unnecessary, same as PrimalSurge and NidaleeE/WBuff.

File: Content/LeagueSandbox-Scripts/Champions/Nidalee/TransformedE.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
+﻿using GameServerCore.Domain;
+using GameServerCore.Domain.GameObjects;
+using LeagueSandbox.GameServer.Scripting.CSharp;
+
+namespace Spells
+{
+    class Swipe : IGameScript
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/Champions/Nidalee/TransformedE.cs]: Unnecessary, same as the other Nidalee additions

File: Content/LeagueSandbox-Scripts/Champions/Nidalee/TransformedQ.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
+﻿using GameServerCore.Domain;
+using GameServerCore.Domain.GameObjects;
+using LeagueSandbox.GameServer.Scripting.CSharp;
+
+namespace Spells
+{
+    class Takedown : IGameScript
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/Champions/Nidalee/TransformedQ.cs]: Unnecessary, same as the other Nidalee additions.

File: Content/LeagueSandbox-Scripts/Champions/Nidalee/TransformedW.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
+﻿using GameServerCore.Domain;
+using GameServerCore.Domain.GameObjects;
+using LeagueSandbox.GameServer.Scripting.CSharp;
+
+namespace Spells
+{
+    class Pounce : IGameScript
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/Champions/Nidalee/TransformedW.cs]: Unnecessary, same as the other Nidalee additions.

File: Content/LeagueSandbox-Scripts/Champions/Nidalee/W.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+using GameServerCore.Domain.GameObjects;
+using GameServerCore.Enums;
+using LeagueSandbox.GameServer.API;
+using LeagueSandbox.GameServer.GameObjects.AttackableUnits.AI;
+using LeagueSandbox.GameServer.Scripting.CSharp;
+using static LeagueSandbox.GameServer.API.ApiFunctionManager;
+
+namespace Spells
+{
+    class Bushwhack : IGameScript
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/Champions/Nidalee/W.cs]: Unnecessary, same as the other Nidalee additions.

File: Content/LeagueSandbox-Scripts/LeagueSandbox-Scripts.csproj [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    <Compile Include="Champions\Karthus\R.cs" />
+    <Compile Include="Champions\Kassadin\Q.cs" />
+    <Compile Include="Champions\Lucian\E.cs" />
+    <Compile Include="Champions\Lucian\Q.cs" />
+    <Compile Include="Champions\Lucian\W.cs" />
+    <Compile Include="Champions\Lulu\R.cs" />
+    <Compile Include="Champions\Lulu\W.cs" />
+    <Compile Include="Champions\Lux\R.cs" />
+    <Compile Include="Champions\MasterYi\R.cs" />
+    <Compile Include="Champions\Nidalee\E.cs" />
```
Comment (microsoftv) [inline Content/LeagueSandbox-Scripts/LeagueSandbox-Scripts.csproj]: Remove unnecessaries from scripts project.

File: GameServerCore/Domain/ICharData.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
+﻿using GameServerCore.Enums;
+
+namespace LeagueSandbox.GameServer.Content
+{
+    public interface ICharData
```
Comment (microsoftv) [inline GameServerCore/Domain/ICharData.cs]: Not sure how to feel about a duplicate CharData lying around in Core/Domain.

File: GameServerLib/Config.cs [hunk 11]
```diff
@@ -72,35 +72,17 @@ private void LoadConfig(Game game, string json)
...
             var gameToken = data.SelectToken("game");
             GameConfig = new GameConfig(gameToken);
 
-            // Read spawns info
-            ContentManager = ContentManager.LoadGameMode(game, GameConfig.GameMode, ContentPath);
-            var mapPath = ContentManager.GetMapDataPath(GameConfig.Map);
-            var mapData = JObject.Parse(File.ReadAllText(mapPath));
-            var spawns = mapData.SelectToken("spawns");
+            // Load data package
+            ContentManager = ContentManager.LoadDataPackage(game, GameConfig.DataPackage, ContentPath);
```
Comment (microsoftv) [inline GameServerLib/Config.cs]: May want to re-add some form of distinction between packages and modes (for another PR most likely).

File: GameServerLib/Content/Package.cs [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+
+namespace LeagueSandbox.GameServer.Content
+{
+    public class Package : IPackage
+    {
+        public string PackagePath { get; private set; }
+        public string PackageName { get; private set; }
+
+        private readonly Dictionary<string, ISpellData> _spellData = new Dictionary<string, ISpellData>();
+        private readonly Dictionary<string, CharData> _charData = new Dictionary<string, CharData>();
```
Comment (microsoftv) [inline GameServerLib/Content/Package.cs]: Use ICharData.
```suggestion
        private readonly Dictionary<string, ICharData> _charData = new Dictionary<string, ICharData>();
```

File: GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs [hunk 13]
```diff
@@ -59,7 +59,9 @@ public class ObjAiBase : AttackableUnit, IObjAiBase
         {
             _itemManager = game.ItemManager;
             _scriptEngine = game.ScriptEngine;
-            CharData = _game.Config.ContentManager.GetCharData(Model);
+
+            CharData = (CharData) _game.Config.ContentManager.GetCharData(Model);
```
Comment (microsoftv) [inline GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs]: Use ICharData.
```suggestion
            CharData = (ICharData) _game.Config.ContentManager.GetCharData(Model);
```
**Requires**:
```
        public ICharData CharData { get; } // line 38 ObjAIBase.cs
```
```
        public void LoadStats(ICharData charData) // line 109 Stats.cs
```

**Review by lzardy — CHANGES_REQUESTED** [review]
Apply changes based on my comments (besides those that are stated as suggestions).

**Comment by abosma** [comment]
Implemented changes requested

**Review by lzardy — APPROVED** [review]
Click resolve on all comments and you should be good to go.

**Comment by lzardy** [comment]
Seems like you'll need to rebase to get the checks to run correctly.

**Comment by abosma** [comment]
Rebased and commited the merge with the latest auto attack change.

**Comment by abosma** [comment]
Removed LeagueSandbox-Scripts from PR, going to make another pull request for adding LeagueSandbox-Scripts and discuss there.

**Review by moonshadow565 — APPROVED** [review]

**Pull request merged**